### PR TITLE
Feature/more improvements

### PR DESCRIPTION
**Pull request by dizzzz**
Another round of sunday-eve code improvements. Grouped per theme.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/protocolhandler/xmldb/XmldbURL.java [hunk 1]
```diff
@@ -280,7 +280,7 @@ public String getProtocol(){
      */
     public String getHost(){
         final String hostname=myUrl.getHost();
-        if("".equals(hostname)){
+        if(hostname != null && hostname.isEmpty()){
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/protocolhandler/xmldb/XmldbURL.java:283]: This isn't the same logic as before!

File: exist-core/src/main/java/org/exist/xmldb/FullXmldbURI.java [hunk 2]
```diff
@@ -188,7 +188,7 @@ private void setContext(String context) {
                     context = context.substring(0, context.length() - 1);
                 }
             }
-            this.context = "".equals(context) ? null : context;
+            this.context = context != null && context.isEmpty() ? null : context;
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xmldb/FullXmldbURI.java:191]: This is not the same logic as before.

**Review by adamretter — COMMENTED** [review]
I am not sure about this pattern of changes:

Before -
```
return groups.toArray(new String[groups.size()]);
```

After:
```
return groups.toArray(new String[0]);
```

As I understand it, after your change two arrays have two be allocated (and then the first is discarded). Whereas before your change only a single array has to be allocated, so I would think that is more efficient.

----

**Comment by dizzzz** [comment]
@adamretter yes I understand: this is the rationale as provided by intellij:

"There are two styles to convert a collection to an array: either using a pre-sized array (like c.toArray(new String[c.size()])) or using an empty array (like c.toArray(new String[0]).

In older Java versions using pre-sized array was recommended, as the reflection call which is necessary to create an array of proper size was quite slow. However since late updates of OpenJDK 6 this call was intrinsified, making the performance of the empty array version the same and sometimes even better, compared to the pre-sized version. Also passing pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray call which may result in extra nulls at the end of the array, if the collection was concurrently shrunk during the operation.

This inspection allows to follow the uniform style: either using an empty array (which is recommended in modern Java) or using a pre-sized array (which might be faster in older Java versions or non-HotSpot based JVMs)."

**Comment by adamretter** [comment]
@dizzzz Hmmmm. I would be surprised if they are mistaken. However reading the code of the JDK doesn't make it seem that way at all.

**Comment by dizzzz** [comment]
Hmmmmm thnx for checking. I assumed indeed that these guys know what is happening under the hood. Maybe it is a compiler optimization step ?

**Comment by adamretter** [comment]
@dizzzz Indeed it could be.

**Pull request merged**